### PR TITLE
NoSuchCacheManagerFailureAnalyzer

### PR DESCRIPTION
**Pull request by nosan**
see [gh-13348](https://github.com/spring-projects/spring-boot/issues/13348)

### CODE REVIEW

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/CacheAutoConfiguration.java [hunk 1]
```diff
@@ -52,8 +57,8 @@
  * Cache store can be auto-detected or specified explicitly via configuration.
  *
  * @author Stephane Nicoll
- * @since 1.3.0
  * @see EnableCaching
+ * @since 1.3.0
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/CacheAutoConfiguration.java]: please refrain from doing unrelated changes.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/CacheAutoConfiguration.java [hunk 2]
```diff
@@ -75,8 +80,8 @@ public CacheManagerCustomizers cacheManagerCustomizers(
 
 	@Bean
 	public CacheManagerValidator cacheAutoConfigurationValidator(
-			CacheProperties cacheProperties, ObjectProvider<CacheManager> cacheManager) {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/CacheAutoConfiguration.java]: You can do `ObjectProvider<Map<String,CacheManager>>` instead of injecting the context

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/CacheAutoConfiguration.java [hunk 3]
```diff
@@ -134,4 +142,45 @@ public void afterPropertiesSet() {
 
 	}
 
+	/**
+	 * Exception thrown when {@link org.springframework.cache.CacheManager} implementation
+	 * are not specified.
+	 */
+	static class NoCacheManagerException extends RuntimeException {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/CacheAutoConfiguration.java]: To mirror `NoSuchBeanDefinitionException`, you could call that `NoSuchCacheManagerException`

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/CacheAutoConfiguration.java [hunk 4]
```diff
@@ -134,4 +142,45 @@ public void afterPropertiesSet() {
...
+		}
+
+	}
+
+	/**
+	 * Exception thrown when multiple {@link org.springframework.cache.CacheManager}
+	 * implementations are available with no way to know which implementation should be
+	 * used.
+	 */
+	static class NonUniqueCacheManagerException extends RuntimeException {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/CacheAutoConfiguration.java]: And this one could be `NoUniqueCacheManagerException` extending from the other exception (again to mirror `NoUniqueBeanDefinitionException`)

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/NoCacheManagerFailureAnalyzer.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import org.springframework.boot.diagnostics.FailureAnalysis;
+
+/**
+ * An {@link AbstractFailureAnalyzer} for
+ * {@link CacheAutoConfiguration.NoCacheManagerException}.
+ *
+ * @author Dmytro Nosan
+ * @since 2.1.0
+ */
+class NoCacheManagerFailureAnalyzer
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/cache/NoCacheManagerFailureAnalyzer.java]: It would be nice to merge the `FailureAnalyzer` so that we only have one class (and an instance check to determine the type). This is my main motivation for having a hierarchy in the two exceptions at hand.

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/cache/CacheAutoConfigurationTests.java [hunk 6]
```diff
@@ -1007,6 +1018,16 @@ public CacheResolver cacheResolver() {
 
 	}
 
+	@Configuration
+	static class AdditionalCacheManagerAutoConfiguration {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/cache/CacheAutoConfigurationTests.java]: That's not an auto-configuration and shouldn't be named like that. It shouldn't be added as an auto-config in the test either.

**Comment by nosan** [comment]
@snicoll Frankly speaking, I really like these  two analyzers  `NoSuchBeanDefinitionFailureAnalyzer` and `NoUniqueBeanDefinitionFailureAnalyzer` and I'd like to reuse them as much as possible, that's why I'd prefer to extend `NoUniqueBeanDefinitionException` and `NoUniqueBeanDefinitionException` accordingly and provide `NoSuchCacheManagerFailureAnalyzer` only for a `spring.cache.type=generic`. 

Thoughts ?

**Comment by nosan** [comment]
I need to revise this PR.

**Comment by nosan** [comment]
@snicoll I've updated this PR. Please take a look :)
Thanks in advance.

**Comment by nosan** [comment]
@snicoll 

Sorry for the long delay, I was a bit busy :(

I've updated this PR and fixed all conflicts.

**Comment by snicoll** [comment]
@nosan no need to apologize, I am the one who didn't get a chance to have a deeper look at it yet.

**Comment by snicoll** [comment]
Thanks for your patience @nosan. I am not keen to bring this amount of complexity for the task at hand. I've created [an issue in Framework](https://github.com/spring-projects/spring-framework/issues/22442) to improve the exception that is throw in such cases so that we can reuse the existing analyzers or very easily tune a new one for our purposes. Thanks for the PR, in any case.

**Comment by nosan** [comment]
@snicoll fair enough, this PR contained a lot of changes.
